src/lib/mongoBackend/MongoCommonUpdate.cpp

```diff
@@ -736,22 +629,39 @@ static bool processLocation(ContextAttributeVector caV, std::string& locAttr, do
 * Returns true if an attribute was deleted, false otherwise
 *
 */
-static bool deleteAttribute(BSONObj& attrs, BSONObj& newAttrs, ContextAttribute* caP) {
-    BSONArrayBuilder newAttrsBuilder;
-    bool deleted = false;
-    for( BSONObj::iterator i = attrs.begin(); i.more(); ) {
-
-        BSONObjBuilder newAttr;
-        if (checkAndDelete(&newAttr, i.next().embeddedObject(), *caP) && !deleted) {
-            deleted = true;
-        }
-        else {
-            newAttrsBuilder.append(newAttr.obj());
-        }
+static bool deleteAttribute(BSONObj& attrs, BSONObjBuilder* toUnset, std::map<std::string, unsigned int>* deleted, ContextAttribute* caP)
+{
+
+  /* Attributes with metadata ID are stored as <attrName>_<ID> in the attributes embedded document */
+  std::string effectiveName = caP->name;
+  if (caP->getId() != "")
+  {
+    effectiveName += "__" + caP->getId();
+  }
+
+  if (attrs.hasField(effectiveName.c_str()))
+  {
+    const std::string composedName = std::string(ENT_ATTRS) + "." + effectiveName;
+    toUnset->append(composedName, 1);
+
+    /* Record the ocurrence of this attribute. Only attributes using ID may have a value greater than 1 */
+    if (deleted->count(caP->name))
+    {
+      std::map<std::string, unsigned int>::iterator it = deleted->find(caP->name);
+      it->second++;
     }
-    newAttrs = newAttrsBuilder.arr();
+    else
+    {
+      deleted->insert(std::make_pair(caP->name, 1));
+    }
+
+    return true;
+  }
+  else
```

Comment:
Like before, 4 lines => 1 line, or better. check and return false before you start